src/trace/implementations/merge_batcher.rs

```diff
@@ -179,81 +143,235 @@ impl<D: Ord, T: Ord, R: Semigroup> MergeSorter<D, T, R> {
+
+                if all {
+                    // All data is ready, push what we accumulated, stash whole block.
+                    if !ship.is_empty() {
+                        block_list.push((Antichain::new(), std::mem::replace(&mut ship, self.empty())));
+                    }
+                    block_list.push((Antichain::new(), block));
+                } else if any {
+                    // Iterate block, sorting items into ship and keep
+                    for datum in block.drain(..) {
+                        if upper.less_equal(&datum.1) {
+                            frontier.insert_ref(&datum.1);
+                            keep_frontier.insert_ref(&datum.1);
+                            keep.push(datum);
+                            if keep.capacity() == keep.len() {
+                                // remember keep
+                                keep_list.push((std::mem::take(&mut keep_frontier), std::mem::replace(&mut keep, self.empty())));
+                            }
+                        } else {
+                            ship.push(datum);
+                            if ship.capacity() == ship.len() {
+                                // Ship is full, push in on the block list, get an empty one.
+                                block_list.push((Antichain::new(), std::mem::replace(&mut ship, self.empty())));
+                            }
+                        }
+                    }
+                    // Recycle leftovers
+                    self.recycle(block);
+                } else {
+                    // Keep the entire block.
+                    if !keep.is_empty() {
+                        keep_list.push((std::mem::take(&mut keep_frontier), std::mem::replace(&mut keep, self.empty())));
+                    }
+                    keep_list.push((block_frontier, block));
+                }
+            }
+
+            // Capture any residue left after iterating blocks.
+            if !ship.is_empty() {
+                block_list.push((Antichain::new(), std::mem::replace(&mut ship, self.empty())));
+            }
+            if !keep.is_empty() {
+                keep_list.push((std::mem::take(&mut keep_frontier), std::mem::replace(&mut keep, self.empty())));
+            }
+
+            // Collect finished chains
+            if !block_list.is_empty() {
+                ship_list.push(block_list);
+            }
+            if !keep_list.is_empty() {
+                self.queue.push(keep_list);
+            }
+        }
+
+        self.account(self.queue.iter().flatten(), 1);
+
+        if ship_list.len() > 0 {
+            self.maintain();
         }
 
-        if let Some(mut last) = self.queue_pop() {
-            ::std::mem::swap(&mut last, target);
+        while ship_list.len() > 1 {
+            let list1 = ship_list.pop().unwrap();
+            let list2 = ship_list.pop().unwrap();
+            ship_list.push(self.merge_by(list1, list2));
```

Comment:
No reason these need to be in the same geometric order as the held updates. There is the opportunity to resort them, or to use a binary heap to continually merge the smallest chains. No strong feelings, because it seems unlikely to be adversarially selected.